### PR TITLE
MAINT: More informative TypeError for IndexEngine.get_loc

### LINKED ISSUE #12218: Empty TypeError message with bad column selection
I was trying to guess, in a dataframe with multi-indexed columns, how to select all columns with a given second-level value.

And then I hit an empty `TypeError` message

```
In [8]: df = pd.DataFrame(columns=['i', 'c', 'x', 'y'], data=[[0, 0, 1, 2], [1, 0, 3, 4], [0, 1, 1, 2], [1, 1, 3, 4]])

In [9]: dg
Out[9]: 
   x     y   
c  0  1  0  1
i            
0  1  1  2  2
1  3  3  4  4

In [12]: dg = df.pivot_table(index='i', columns='c', values=['x', 'y'])

n [19]: dg['x', 0]
Out[19]: 
i
0    1
1    3
Name: (x, 0), dtype: int64

In [20]: dg[:, 0]
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-20-7eb22191791d> in <module>()
----> 1 dg[:, 0]

/home/nicolas/Git/pandas/pandas/core/frame.py in __getitem__(self, key)
   1977             return self._getitem_frame(key)
   1978         elif is_mi_columns:
-> 1979             return self._getitem_multilevel(key)
   1980         else:
   1981             return self._getitem_column(key)

/home/nicolas/Git/pandas/pandas/core/frame.py in _getitem_multilevel(self, key)
   2021 
   2022     def _getitem_multilevel(self, key):
-> 2023         loc = self.columns.get_loc(key)
   2024         if isinstance(loc, (slice, Series, np.ndarray, Index)):
   2025             new_columns = self.columns[loc]

/home/nicolas/Git/pandas/pandas/indexes/multi.pyc in get_loc(self, key, method)
   1526             key = _values_from_object(key)
   1527             key = tuple(map(_maybe_str_to_time_stamp, key, self.levels))
-> 1528             return self._engine.get_loc(key)
   1529 
   1530         # -- partial selection or non-unique index

/home/nicolas/Git/pandas/pandas/index.pyx in pandas.index.IndexEngine.get_loc (pandas/index.c:4005)()
    135             util.set_value_at(arr, loc, value)
    136 
--> 137     cpdef get_loc(self, object val):
    138         if is_definitely_invalid_key(val):
    139             raise TypeError

/home/nicolas/Git/pandas/pandas/index.pyx in pandas.index.IndexEngine.get_loc (pandas/index.c:3617)()
    137     cpdef get_loc(self, object val):
    138         if is_definitely_invalid_key(val):
--> 139             raise TypeError
    140 
    141         if self.over_size_threshold and self.is_monotonic_increasing:

TypeError: 
```

### PR DESCRIPTION
**Pull request by gfyoung**
Title is self-explanatory.  Closes #12218.

### CODE REVIEW

**Comment by gfyoung** [comment]
Are either of the following necessary for this PR:

1) What's New (if yes, which section?)
2) Unit test (if yes, where?)

**Comment by jreback** [comment]
I don't believe this ever becomes an external exception

**Comment by gfyoung** [comment]
But the issue that I'm closing showed an example where the Exception was thrown

**Comment by jreback** [comment]
so this is right

so add a test that asserts the contents of the error message

it's s type error because the indexer is not the correct type to select the level

**Comment by gfyoung** [comment]
Added a test, and the test passes on Travis.  Should be good to merge if there is nothing else.

File: pandas/tests/frame/test_indexing.py [hunk 1]
```diff
@@ -2590,3 +2590,19 @@ def test_head_tail(self):
         empty_df = DataFrame()
         assert_frame_equal(empty_df.tail(), empty_df)
         assert_frame_equal(empty_df.head(), empty_df)
+
+    def test_type_error_multiindex(self):
+        df = pd.DataFrame(columns=['i', 'c', 'x', 'y'],
```
Comment (jreback) [inline pandas/tests/frame/test_indexing.py]: add the issue number here

File: pandas/tests/frame/test_indexing.py [hunk 2]
```diff
@@ -2590,3 +2590,19 @@ def test_head_tail(self):
...
+                          data=[[0, 0, 1, 2], [1, 0, 3, 4],
+                                [0, 1, 1, 2], [1, 1, 3, 4]])
+        dg = df.pivot_table(index='i', columns='c',
+                            values=['x', 'y'])
+
+        try:
+            print(dg[:, 0])
```
Comment (jreback) [inline pandas/tests/frame/test_indexing.py]: you can do ``str(dg[:, 0])`` to trigger an error and use an ``.assertRaises`` to test for it.

File: pandas/tests/frame/test_indexing.py [hunk 3]
```diff
@@ -2590,3 +2590,19 @@ def test_head_tail(self):
         empty_df = DataFrame()
         assert_frame_equal(empty_df.tail(), empty_df)
         assert_frame_equal(empty_df.head(), empty_df)
+
+    def test_type_error_multiindex(self):
+        df = pd.DataFrame(columns=['i', 'c', 'x', 'y'],
+                          data=[[0, 0, 1, 2], [1, 0, 3, 4],
+                                [0, 1, 1, 2], [1, 1, 3, 4]])
+        dg = df.pivot_table(index='i', columns='c',
+                            values=['x', 'y'])
```
Comment (jreback) [inline pandas/tests/frame/test_indexing.py]: futher use the other part of the test in the issue to assert that ``dg['x', 0]`` (and ``dg[('x', 0), :]`` works)

File: pandas/tests/frame/test_indexing.py (hunk 1 above)
Comment (gfyoung) [inline pandas/tests/frame/test_indexing.py]: Done.

File: pandas/tests/frame/test_indexing.py (hunk 3 above)
Comment (gfyoung) [inline pandas/tests/frame/test_indexing.py]: Done.

Comment (gfyoung) [inline pandas/tests/frame/test_indexing.py]: Done.

File: pandas/tests/frame/test_indexing.py (hunk 2 above)
Comment (gfyoung) [inline pandas/tests/frame/test_indexing.py]: Used `assertRaisesRegexp` instead so I could also test for the error message.

**Comment by gfyoung** [comment]
Not really sure why Python 3.4 is failing, as the failing test case has no relation to what I changed.  I'm going to rebase onto `master` so as to trigger Travis again.

File: pandas/tests/frame/test_indexing.py [hunk 4]
```diff
@@ -2590,3 +2590,24 @@ def test_head_tail(self):
...
+
+        with assertRaisesRegexp(TypeError, invalid_key):
+            str(dg[:, 0])
+
+        # Neither of the following commands should
+        # throw a TypeError because they are valid
```
Comment (jreback) [inline pandas/tests/frame/test_indexing.py]: you don't need to put these asserts around here, just run the actual commands. if they raise the tests will fail; this just obfuscuates tings.  best is actually to test the results

e.g.

```
result = dg.loc[:, (slice(None), 0)]
expected = ....
assert_.....
```

Comment (gfyoung) [inline pandas/tests/frame/test_indexing.py]: Done.

**Comment by gfyoung** [comment]
Tests are passing, so this should be good to merge if there is nothing else.

**Comment by jreback** [comment]
can you add a whatsnew note (you can put in bug fixes is fine). ping when pushed.

**Comment by gfyoung** [comment]
Should I add it to v0.18.0 or v0.18.1?

**Comment by jreback** [comment]
018.0 is ok

**Comment by gfyoung** [comment]
Sorry, as soon as I posted the question, I saw you had already added it to the `0.18.0` milestone, which made the answer somewhat obvious. :smile:

File: doc/source/whatsnew/v0.18.0.txt [hunk 5]
```diff
@@ -1117,3 +1117,4 @@ Bug Fixes
 - Bug in ``DataFrame.apply`` in which reduction was not being prevented for cases in which ``dtype`` was not a numpy dtype (:issue:`12244`)
 - Bug when initializing categorical series with a scalar value. (:issue:`12336`)
 - Bug when specifying a UTC ``DatetimeIndex`` by setting ``utc=True`` in ``.to_datetime`` (:issue:11934)
+- Bug in ``IndexEngine.get_loc`` in which an empty ``TypeError`` was being thrown instead of one with an error message (:issue:12218)
```
Comment (jreback) [inline doc/source/whatsnew/v0.18.0.txt]: just say that you added a meaningful error message when multi-axis indexing with an invalid value.

these whatsnotes are for the average user. Those who want/need more detail can simply click on the issue itself. But for a quick glance, just need to highlite terms that most users would know.

File: pandas/tests/frame/test_indexing.py [hunk 6]
```diff
@@ -2590,3 +2590,32 @@ def test_head_tail(self):
...
+                            values=['x', 'y'])
+
+        with assertRaisesRegexp(TypeError, "is an invalid key"):
+            str(dg[:, 0])
+
+        columns = pd.MultiIndex(levels=[['x', 'y'], [0, 1]],
+                                labels=[[0, 1], [0, 0]],
+                                names=[None, 'c'])
+        values = np.array([[1, 2], [3, 4]], dtype=np.int64)
+        index = pd.Int64Index(data=[0, 1], dtype='int64', name='i')
```
Comment (jreback) [inline pandas/tests/frame/test_indexing.py]: ``Index(range(2), name='i')``

File: pandas/tests/frame/test_indexing.py [hunk 7]
```diff
@@ -2590,3 +2590,32 @@ def test_head_tail(self):
...
+        expected = pd.DataFrame(data=values, columns=columns, index=index)
+
+        result = dg.loc[:, (slice(None), 0)]
+        assert_frame_equal(result, expected)
+
+        name = ('x', 0)
+        values = np.array([1, 3], dtype=np.int64)
+        index = pd.Int64Index(data=[0, 1], dtype='int64', name='i')
```
Comment (jreback) [inline pandas/tests/frame/test_indexing.py]: same here, leave off unless its germane (which its not as its a regular index)

File: pandas/tests/frame/test_indexing.py (hunk 6 above)
Comment (gfyoung) [inline pandas/tests/frame/test_indexing.py]: Sorry, what do you mean?

File: doc/source/whatsnew/v0.18.0.txt (hunk 5 above)
Comment (gfyoung) [inline doc/source/whatsnew/v0.18.0.txt]: Ah, okay.  Good to know.

File: pandas/tests/frame/test_indexing.py (hunk 7 above)
Comment (gfyoung) [inline pandas/tests/frame/test_indexing.py]: Could you clarify your comment?

File: pandas/tests/frame/test_indexing.py (hunk 6 above)
Comment (jreback) [inline pandas/tests/frame/test_indexing.py]: you dont need the index to the expected ``DataFrame`` constructor, add it just obfuscates what you are doing. further don't use ``pd.`` everything is already imported

File: pandas/tests/frame/test_indexing.py (hunk 7 above)
Comment (jreback) [inline pandas/tests/frame/test_indexing.py]: same as above, you don't need to explicit create an index unless its germane to the question which its not.

File: pandas/tests/frame/test_indexing.py (hunk 6 above)
Comment (gfyoung) [inline pandas/tests/frame/test_indexing.py]: Unfortunately, the `DataFrame` objects no longer match if you omit the `index` argument to the constructor.  However, I do agree that specifying an `Int64Index` was not necessary.

Comment (jreback) [inline pandas/tests/frame/test_indexing.py]: pls show an example then

Comment (gfyoung) [inline pandas/tests/frame/test_indexing.py]: You mean include a test where the `index` argument is not needed?

File: pandas/tests/frame/test_indexing.py [hunk 8]
```diff
@@ -2590,3 +2590,32 @@ def test_head_tail(self):
...
+                            values=['x', 'y'])
+
+        with assertRaisesRegexp(TypeError, "is an invalid key"):
+            str(dg[:, 0])
+
+        columns = MultiIndex(levels=[['x', 'y'], [0, 1]],
+                             labels=[[0, 1], [0, 0]],
+                             names=[None, 'c'])
+        values = np.array([[1, 2], [3, 4]], dtype=np.int64)
+        index = Index(data=[0, 1], dtype='int64', name='i')
```
Comment (jreback) [inline pandas/tests/frame/test_indexing.py]: oh I see you the index has a name, ok then.

Still construct like this 

``index = Index(range(2), name='i')`` it automatically is int64
you can just construct 
```
expected = DataFrame([[1, 2], [3, 4], columns=columns, dtype='int64')
```

Comment (jreback) [inline pandas/tests/frame/test_indexing.py]: no need to intermix numpy arrays at all here. Its just confusing to a reader.

Comment (gfyoung) [inline pandas/tests/frame/test_indexing.py]: Sure thing.  Done.

**Comment by gfyoung** [comment]
Could you cancel some of my duplicate, older builds?  I see no reason for them to run given all of the updating I have done.

**Comment by jreback** [comment]
done, do you know any way to configure that on Travis? (e.g. it builds pull-requests, but I want force pushes to cancel anything but the last one).

**Comment by gfyoung** [comment]
Unfortunately not.  See <a href="https://github.com/travis-ci/travis-ci/issues/964">here</a>

**Comment by jreback** [comment]
actually I *think* this will work as the ref commit is now gone. (if you force push on the same branch)

**Comment by gfyoung** [comment]
I'm not entirely sure I follow what you mean since I do force pushes all time with my PR's in `pandas` and haven't seen that happen.

**Comment by jreback** [comment]
what I mean is that if you force push, then when Travis tries to build it is fails (but is grey so doesn't count) and doesn't build the previous commit (but instead builds your new one)

**Comment by gfyoung** [comment]
On a separate note, is there a `pandas-svn` mailing list (similar to `numpy` and `scipy`) that GitHub can send notifications to to let us know when merges / commits have been made to the `master` branch?  I find such notifications to be quite useful so that PR branches don't get too far behind the `master` branch.

**Comment by gfyoung** [comment]
Regarding Travis: that should work I suppose, but I haven't seen that work consistently in practice, at least not in this library.

**Comment by jreback** [comment]
@gfyoung where is this list defined?

yes easy to add to Travis

**Comment by gfyoung** [comment]
Oh, I meant : does such a mailing list exist?

**Comment by jreback** [comment]
@gfyoung could create one, what I meant was does ``numpy`` define one somewhere?

@jorisvandenbossche @TomAugspurger would this be useful?

**Comment by gfyoung** [comment]
@jreback : If you visit this link <a href="http://www.scipy.org/scipylib/mailing-lists.html">here</a>, you'll see it there.  Every time PR's get merged into `master` or commits are made, everyone on that list gets notified.

**Comment by jreback** [comment]
@gfyoung can you make a separate issue about this. (and include that pointer)

**Comment by jreback** [comment]
thanks!

**Pull request closed without merging**